File: random-variable/box_muller.f90
!=================================================================
! BOX-MULLER METHOD
!=================================================================
! Author: Daniel Noal Pineda
! Email : noaldaniel41@gmail.com
! Date  : 2025
! Repository: https://github.com/tuusuario/tu-repo
!=================================================================
! OBJECTIVES: Generate Gaussian random numbers N(0,1) using the 
!             Box-Müller transformation method
!
! INPUTS:
!         ·ndat  : number of data points to generate
!
! OUTPUTS:
!         ·xgaus : array containing the generated Gaussian numbers N(0,1)
!
! NOTES:
!         To generate Gaussian numbers with other constants sigma, mu:
!         N(mu,sigma) = sigma*N(0,1) + mu
!=================================================================

SUBROUTINE SUBGAUSS(ndat,xgaus)

    implicit none

    ! Inputs
    integer, intent(in) :: ndat

    ! Outputs
    double precision, intent(out) :: xgaus(ndat)

    ! Internal Variables
    double precision :: X1, X2, r, theta
    double precision, parameter :: Pi = dacos(-1.0d0)
    integer :: i

    ! BOx-MÜLLER METHOD
    do i=1,ndat,2

        ! Generate 2 random numbers in [0,1]
        X1 = rand() 
        X2 = rand()

        ! Use the variable change of the method
        r = dsqrt(-2.0d0*log(X1))
        theta = 2.0d0*Pi*X2

        ! Obtain two normal numbers N(0,1)
        xgaus(i) = r*dcos(theta)

        ! The second normal number is only stored if possible (if ndat is odd, the last one will be left over)
        if (i+1 .le. ndat) then 
            
            xgaus(i+1) = r*dsin(theta)

        endif

    enddo

    RETURN

END SUBROUTINE SUBGAUSS

File: Python/random-variable/box_muller.py
#=================================================================
# BOX-MULLER METHOD
#=================================================================
# Author: Daniel Noal Pineda
# Email : noaldaniel41@gmail.com
# Date  : 2025
# Repository: https://github.com/Dnoal17/numerical-methods.git
#=================================================================
# OBJECTIVES: Generate Gaussian random numbers N(0,1) using the 
#             Box-Müller transformation method
#
# NOTES: To generate Gaussian numbers with other constants sigma, mu:
#        N(mu,sigma) = sigma*N(0,1) + mu
#
# INPUTS:
#         ·ndat  : number of points to generate
#
# OUTPUTS:
#         ·xgaus : Numoy array containing the generated Gaussian numbers N(0,1)
#=================================================================

import numpy as np

def gauss(ndat):

    # The algorithm need ndat to be even. In case the user sets an odd number one extra point will be generated
    if ndat % 2 != 0:
        m = ndat + 1

    else:
        m = ndat

    # Generate m random numbers in U(0,1)
    u = np.random.rand(m)

    # Determine the parameters of the Box-Muller method, r and theta, and initialize the output array
    r = np.sqrt(-2 * np.log(u[0::2]))
    theta = 2 * np.pi * u[1::2]
    x = np.empty(m)

    # Generate the gaussian numbers N(0,1)
    x[0::2] = r * np.cos(theta)
    x[1::2] = r * np.sin(theta)

    # Return the array with the length the user set
    return x[:ndat]